src/targets/pubDev.ts

```diff
@@ -0,0 +1,208 @@
+    githubRepo: GitHubGlobalConfig
+  ) {
+    super(config, artifactProvider, githubRepo);
+    this.pubDevConfig = this.getPubDevConfig();
+    this.checkRequiredSoftware();
+    this.githubRepo = githubRepo;
+  }
+
+  /**
+   * Returns the pub-dev config with the required data (e.g. environment
+   * variables) for this target. If there's a configuration requirement missing,
+   * raises an error.
+   *
+   * @returns the pub-dev config for this target.
+   */
+  private getPubDevConfig(): PubDevTargetConfig {
+    // We could do `...this.config`, but `packages` is in a list, not array format in `.yml`
+    // so I wanted to keep setting the defaults unified.
+    return {
+      dartCliPath: this.config.dartCliPath || 'dart',
+      packages: this.config.packages
+        ? Object.keys(this.config.packages)
+        : ['.'],
+      ...this.getTargetSecrets(),
+    };
+  }
+
+  private getTargetSecrets(): Record<SecretsType, string> {
+    const secrets = targetSecrets.map(name => {
+      checkEnvForPrerequisite({ name });
+      return {
+        name,
+        value: process.env[name],
+      };
+    });
+    return this.reduceConfig(secrets);
+  }
+
+  private reduceConfig(config: PartialTargetConfig): Record<string, string> {
+    return config.reduce((prev, current) => {
+      return {
+        ...prev,
+        [current.name]: current.value,
+      };
+    }, {});
+  }
+
+  /**
+   * Checks whether the required software to run this target is available
+   * in the system. It assumes the config for this target to be available.
+   * If there's required software missing, raises an error.
+   */
+  private checkRequiredSoftware(): void {
+    this.logger.debug(
+      'Checking if Dart CLI is available: ',
+      this.pubDevConfig.dartCliPath
+    );
+    checkExecutableIsPresent(this.pubDevConfig.dartCliPath);
+  }
+
+  /**
+   * Uploads all files to pub.dev using Dart CLI.
+   *
+   * @param version New version to be released
+   * @param revision Git commit SHA to be published
+   */
+  public async publish(_version: string, revision: string): Promise<any> {
+    await this.createCredentialsFile();
+
+    await withTempDir(
+      async directory => {
+        await this.cloneRepository(this.githubRepo, revision, directory);
+        await forEachChained(this.pubDevConfig.packages, async pkg =>
+          this.publishPackage(directory, pkg)
+        );
+      },
+      true,
+      'craft-pub-dev-'
+    );
+  }
+
+  public async createCredentialsFile(): Promise<void> {
+    const credentialsFilePath = this.getCredentialsFilePath();
+    const content = {
+      accessToken: this.pubDevConfig.PUBDEV_ACCESS_TOKEN,
+      refreshToken: this.pubDevConfig.PUBDEV_REFRESH_TOKEN,
+      tokenEndpoint: 'https://accounts.google.com/o/oauth2/token',
+      scopes: ['openid', 'https://www.googleapis.com/auth/userinfo.email'],
+      expiration: 1645564942000, // Expiration date is required, but irrelevant
```

Comment:
Is the expiration date irrelevant even if it is in the past?

Reply:
I hope so, as it is based on other actions in the wild - https://github.com/k-paxian/dart-package-publisher/blob/519ee1a18f2c9d50534dd722e71275e1ecd67e00/entrypoint.sh#L125-L131

Refresh token should do the job here, but if that'll be restricted, we'll have to think of other solutions.